Make kubelet never delete files on mounted filesystems

jsafrane opened the pull request:
With bug #27653, kubelet could remove mounted volumes and delete user data.
The bug itself is fixed, however our trust in kubelet is significantly lower.
Let's add an extra version of RemoveAll that does not cross mount boundary
(rm -rf --one-file-system).

It calls lstat(path) three times for each removed directory - once in
RemoveAllOneFilesystem and twice in IsLikelyNotMountPoint, however this way
it's platform independent and the directory that is being removed by kubelet
should be almost empty.

k8s-oncall commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/37698)
<!-- Reviewable:end -->

jsafrane commented:
Marked as WIP for dicsussion and it needs a test (mounting in tests could be difficult though).

@saad-ali, @jingxu97, would you consider this patch useful as a safe-guard in kubelet? Will it break anything?

@kubernetes/rh-storage PTAL. Tests are welcome.

rootfs commented on pkg/util/util.go:
```diff
@@ -129,3 +133,85 @@ func Int32PtrDerefOr(ptr *int32, def int32) int32 {
+// files from another filesystems. Like 'rm -rf --one-file-system'.
+// It is copied from RemoveAll() sources, with IsLikelyNotMountPoint
+func RemoveAllOneFilesystem(mounter mount.Interface, path string) error {
+	// Simple case: if Remove works, we're done.
```
check mountpoint here, os.Remove can potentially remove dir too
https://golang.org/pkg/os/#Remove

eparis replied:
os.Remove() will only remove an empty directory. So it shouldn't work if this is a mountpoint or if it has files in in.

wongma7 commented on pkg/util/util.go:
```diff
@@ -129,3 +133,85 @@ func Int32PtrDerefOr(ptr *int32, def int32) int32 {
+	// Remove contents & return first error.
+	err = nil
+	for {
+		names, err1 := fd.Readdirnames(100)
```
why 100, not -1?

rootfs replied:
@wongma7 that's a question for [os.RemoveAll](https://golang.org/src/os/path.go)

wongma7 replied:
oh, it's reading 100 at a time in the loop.

jsafrane commented:
Unit test added

jsafrane commented:
Tested manually with Kubernetes 1.3 with #27653 still reproducible - it does not remove anything in mounted secret (tmpfs) and an AWS volume and shows in the log:

```
Failed to remove orphaned pod "869411ac-b7ce-11e6-90d8-0e2ed4bd5858" dir; err: cannot delete directory /var/lib/origin/openshift.local.volumes/pods/869411ac-b7ce-11e6-90d8-0e2ed4bd5858/volumes/kubernetes.io~secret/default-token-9rqfl: it is a mount point
```

The secret and my AWS volume are still attached and mounted and they're never unmounted (which is fixed in #27653, this PR is just safety check).

jsafrane commented on pkg/util/util.go:
```diff
@@ -129,3 +133,85 @@ func Int32PtrDerefOr(ptr *int32, def int32) int32 {
+	}
+
+	// Close directory, because windows won't remove opened directory.
+	fd.Close()
```
that's a question for  [os.RemoveAll](https://golang.org/src/os/path.go)

jingxu97 commented:
@jsafrane The main concern of this PR is that adding the os.stat() function in removing directories. In current design, cleanupOrphanedPodDirs is called directly by the main kubelet sync loop, such calls might cause hang problem for cleanup NFS mounted directories. I am not sure if we should consider using separate thread to do the cleanup.

jsafrane commented:
@jingxu97, `os.RemoveAll` calls `os.Stat()` too, this PR does not change it. It adds more stat calls, however if it gets stuck, `os.RemoveAll` would get stuck too.

jingxu97 commented:
@jsafrane I see. Another thing I want to check is that before calling RemoveAll() directory, there is a check to see whether there are any volume directories exist. It looks to me this is a strong check already. As long as there is a volume path exist, it will return without remove path. I have an old PR #33616 which tries to check whether there is any mounted directories inside volumes path. Only if the volume path is still mounted, it will skip remove path. Later I removed this logic in #37255 because users complain about NFS hung problem. But as you said, since RemoveAll also call os.Stat(), this check will not add anything worse. 
So I think we only need to put this IsLikelyNotMounted check in volume directory instead of checking it in every sub directory of the pod, which may cause delay for cleanup pod routine. 
I also think we might be able to use a separate goroutine to do this cleanup. This issue was discussed before #13418

jsafrane commented:
I think there is not much in the pod directory - there are `volumes/` and `plugins/` (both with mounted stuff) and `containers/` which just holds empty files. If everything is unmounted, there are just three empty directories to check. If something is mounted, we find it really quickly. Beauty of this `RemoveAllOneFilesystem` is that it does not depend on implementation details. If we restructure the pod directory in the future, this will keep working and it will prevent data being deleted.

jingxu97 commented:
@jsafrane I see. Do you think we should remove the check whether there are directories exist under /volumes? Because I think that check is too strict and might fail to clean up orphaned pods.

jsafrane commented:
`@jingxu97, there already is a check for existence of `volumes/*`, just a few lines above in `getPodVolumePathListFromDisk`. This `RemoveAllOneFilesystem` is another safety mechanism to be 1000% sure that we don't remove any data on mounted filesystems if all the checks above fail or are buggy.

jingxu97 commented on pkg/kubelet/kubelet_volumes.go:
```diff
@@ -111,7 +111,7 @@ func (kl *Kubelet) cleanupOrphanedPodDirs(
 			continue
 		}
```
I think the we should remove the check len(volumePaths) > 0. If we have this check, any directories exist under /volumes will skip the clean up and  RemoveAllOneFilesystem will not be called. The purpose of checking len(volumePaths) > 0 was to prevent any data loss in case the directories under /volumes are not cleaned up properly. Currently, it is possible that the directories under /volumes are unmounted, but fail to delete. If this happens, cleanupOrphanedPod will fail to delete the pod directory.

jingxu97 replied:
Talk with @jsafrane offline, will address this issue in a separate PR. Opened up an issue to track this #41930

jingxu97 commented:
/lgtm

jsafrane commented:
@k8s-bot test this
"Jenkins overloaded. Please try again later"

jsafrane commented:
@k8s-bot kubemark e2e test this
@k8s-bot gce etcd3 e2e test this
@k8s-bot kops aws e2e test this

jsafrane commented:
@thockin, please approve

thockin commented:
It would be nice to have a benchmark on this vs plain old RemoveAll.

rmmh commented:
@k8s-bot verify test this

A [batch test](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/batch/pull-kubernetes-verify/19841/) failed indicating that this needs a staging client-go update.

jsafrane committed:
Make kubelet never delete files on mounted filesystems

With bug #27653, kubelet could remove mounted volumes and delete user data.
The bug itself is fixed, however our trust in kubelet is significantly lower.
Let's add an extra version of RemoveAll that does not cross mount boundary
(rm -rf --one-file-system).

It calls lstat(path) three times for each removed directory - once in
RemoveAllOneFilesystem and twice in IsLikelyNotMountPoint, however this way
it's platform independent and the directory that is being removed by kubelet
should be almost empty.

jsafrane commented:
I moved the function to standalone package at `pkg/util/removeall`.

`pkg/util/util.go` is part of client-go and adding `RemoveAllOneFilesystem` there would pull mounter and exec into client-go as dependencies, which IMO does not belong there.

g0194776 commented:
Will the issue be merge in the v1.6.5 version of Kubernetes?

jsafrane commented:
Well, nobody really asked for backport and I think we missed 1.6.5 deadline. This PR is just a safety mechanism that #27653 won't ever happen again, however the real bugfix is in #27970 and it has been already merged in every branch starting with 1.3.

g0194776 commented:
@jsafrane We are using a cluster of Kubernetes v1.3.4 with 50+ nodes. We've found an issue that the Kubelet will automatically delete all NFS mounted files after the POD became to orphan. This Issue JUST likes what problem we met, so I really care about the BUG FIX about that.

I have no way to prove that the version of v1.6.4 Kubelet won't deletes the mounted files again.
So, do you have any suggestions?

wongma7 commented:
@g0194776 the fix is in 1.3 branch starting only with 1.3.6 https://github.com/kubernetes/kubernetes/releases/tag/v1.3.6. The fix is definitely in 1.6.4

g0194776 commented:
@wongma7 Obviously, It does not fix YET. I was checking the source code of v1.6.4 out on few days ago, I also checked the code implementation at [here.](https://github.com/kubernetes/kubernetes/blob/v1.6.4/pkg/kubelet/kubelet_volumes.go#L118)

The newest change JUST replaces **os.RemoveAll** to another method which called: "removeall.RemoveAllOneFilesystem".

wongma7 commented:
As jsafrane said the real bug fix is #27970
<https://github.com/kubernetes/kubernetes/pull/27970> and that's what I
mean when I say "the fix." If you insist on getting this patch as well then
it needs to be backported because like he said, nobody asked for it until
now. Does that make sense?

On Thu, Jun 22, 2017 at 8:30 AM, Kevin Yang <notifications@github.com>
wrote:

> @wongma7 <https://github.com/wongma7> Obviously, It does not fix YET. I
> was checking the source code of v1.6.4 out on few days ago, I also checked
> the code implementation at here.
> <https://github.com/kubernetes/kubernetes/blob/v1.6.4/pkg/kubelet/kubelet_volumes.go#L118>
>
> The newest change JUST replaces *os.RemoveAll* to another method which
> called: "removeall.RemoveAllOneFilesystem".
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/37698#issuecomment-310365627>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMgP-Isvrtq4Hfe5pQPBknSKaH-CjmW4ks5sGl5VgaJpZM4LAakP>
> .
>

g0194776 commented:
@wongma7  Okay, I got it.
Thanks a lot.
